### ISSUE TITLE
y-axis log scale bug

### ISSUE THREAD

**Issue by jebej** [JuliaPlots/Plots.jl#920]
The y-axis logscale option does not work in PyPlot and GR for log values smaller than or equal to 1:

```julia
heatmap([1,2], [10, 32], rand(2,2), yscale = :log10)
ERROR: DomainError:
 in nan_dom_err at .\math.jl:196 [inlined]
 in log10(::Float64) at .\math.jl:202
 in (::Plots.##114#115{Symbol})(::Float64) at C:\Users\Jeremy\.julia\v0.5\Plots\src\axes.jl:148
...
```

Not sure why this is getting a domain error. Plots must be transforming the y-axis values.

**Comment by jebej** [comment]
This has to do with how Plots calculate the axes limits, which for heatmaps in GR and PyPlot are not the same as the limits of the given y vector.

**Comment by daschw** [comment]
Thanks @jebej for reporting this, but I think the issue a little different. It has nothing to do with the axis (`xaxis` or `yaxis`) or on the values of `z`, but with the vallues of `x`, or `y` respectively, and the fact that you are plotting a heatmap.
If e.g. `x[2] >= 2 * x[1]` on a `log10` `xaxis` the first edge of the heatmap is not positive and `log10` throws a `DomainError`.
You can try this by comparing
```julia
heatmap([1, 3], [1, 2], rand(2,2), xscale = :log10)
```
to
```julia
heatmap([1, 2.99], [1, 2], rand(2,2), xscale = :log10)
```

**Comment by jebej** [comment]
You're right that there are no differences between x and y. However, I do think the issue is in how the limits are calculated. In fact they are calculated differently for plotly (or actually, they are not calculated), which I assume is because heatmaps are supported natively.

**Comment by daschw** [comment]
Yes they are calculated [here](https://github.com/JuliaPlots/Plots.jl/blob/master/src/utils.jl#L336). What would the expected behaviour for a plot on a log axis reaching into its non-positve area be? Should it be cropped at 0 or should there be an error (maybe more verbose about the reason)?

**Comment by jebej** [comment]
I'm not sure what you mean, there are no negative numbers involved here... If the x values passed are valid, ie there are no zeros or negative values, it should work, just like plotly does right now:

![image](https://user-images.githubusercontent.com/11141558/27014811-27ac70ba-4ece-11e7-9e96-aec6c22ed0be.png)

**Comment by jebej** [comment]
Are you asking about how the limits themselves should be calculated? Plotly here does a few things. I believe it is associating each "tile" to an x-axis value, and placing the edge of that tile at that value. This mean there is a tile in between each x-value, and we actually ignore one column of the given matrix. This example might make it a bit clearer:

```julia
heatmap([0.1,1,10], [1, 2], rand(2,2), xscale = :log10)
```
![image](https://user-images.githubusercontent.com/11141558/27014861-27688b24-4ecf-11e7-9225-1626825e37dd.png)

It's debatable that we would want to do that. If a matrix is passed, I believe we should plot the full matrix, and _center_ each tile (matrix entry) on the corresponding (x,y) point.

**Comment by daschw** [comment]
Hm, here are the results of
```julia
using Plots
x = y = 1:10
z = x * y'
heatmap(x, y, z, xscale = :log10, yscale = :log10)
```
for the different backends:
* **GR:**
![heatmap_log10_gr](https://user-images.githubusercontent.com/16589944/27054213-79f3835c-4fc0-11e7-9169-c5945b807b7a.png)

* **PyPlot:**
![heatmap_log10_pyplot](https://user-images.githubusercontent.com/16589944/27054268-af2a9074-4fc0-11e7-8e32-01685d00e45e.png)

* **PlotlyJS:**
![heatmap_log10_plotlyjs](https://user-images.githubusercontent.com/16589944/27055595-3d415a38-4fc5-11e7-88e4-14c5c70f5c20.png)

I think for GR we should try to copy the behaviour of PlotlyJS and PyPlot. Also the colorbar ticks should not have a log scale.

If the `x` and `y` values are changed to `x = y = 1:3:28`, PlotlyJS produces this
![heatmap_log10_plotlyjs_cut](https://user-images.githubusercontent.com/16589944/27056236-9442f100-4fc7-11e7-8bf7-f50048eb2e3b.png)

which looks like the first values of `x` and `y` and the first row and column of `z` are just ignored.
PyPlot and GR throw a `DomainError`, because they would set the first value of the heatmap edges to -1 and pass this to `log10`.
Should the strategy of PlotlyJS also be used for GR and PyPlot?

**Comment by daschw** [comment]
cf. #940

**Comment by yha** [comment]
> Should the strategy of PlotlyJS also be used for GR and PyPlot?

I may be missing something, but I think there is really only one reasonable behavior: `heatmap(x,y,z,xscale=:log10)` should look exactly the same as `heatmap(log10.(x),y,z)` apart from tick marks/labels (and similarly, mutatis mutandis, for the y axis and the `:log2` scale). This means that bin edges should be computed the same way as they are computed for linear scale, except in "log space".
In particular, linear scale `heatmap` centers the ticks on each x/y value, which I think is a good default, so the same should happen in log scale (of course, in log scale, that mean the tick is at the geometric rather than linear average of the bin edges). This way, there is also no problem with non-positive edges as long as the x/y data is positive. And there should be no need to throw away the first row/column just because of using log scale like PlotlyJS does right now.

I would expect the two parts of the following plot to look the same apart from ticks, so the bin edges should be evenly spaced for the second subplot:
```
x = logspace(-1,4,6)
y = logspace(-1,1,5)
z = log10.(x.*y')
p1 = heatmap(log10.(x),log10.(y),z')
p2 = heatmap(x,y,z',xscale=:log10,yscale=:log10)
plot(p1,p2)
````

With GR, the bin edges are completely wrong: most bins don't even contain the relevant x/y values
![log-heatmap-plots grbackend](https://user-images.githubusercontent.com/4170948/41205021-279b5d02-6cf5-11e8-8c97-5000a9e15f22.png)
PyPlot at least gets the original x/y values inside the bins, but the edges are not evenly spaced
![log-heatmap-plots pyplotbackend](https://user-images.githubusercontent.com/4170948/41205031-420a95c2-6cf5-11e8-938a-3b502c702cf9.png)
Plotly/PlotlyJS discards the first row/column and the edges are not evenly spaced
![newplot](https://user-images.githubusercontent.com/4170948/41205032-452f7ce0-6cf5-11e8-966b-5f9a28ed69cd.png)
